src/npy/mod.rs

```diff
@@ -656,12 +739,251 @@ impl ReadableElement for bool {
+pub trait ViewNpyMutExt<'a>: Sized {
+    /// Returns a read-write view of a memory-mapped `.npy` file.
+    ///
+    /// **Note:** Iterates over `bool` array to ensure `0x00`/`0x01` values.
+    fn view_npy_mut(buf: &'a mut [u8]) -> Result<Self, ReadNpyError>;
+}
+
+impl<'a, A, D> ViewNpyExt<'a> for ArrayView<'a, A, D>
+where
+    A: CastableElement,
+    D: Dimension,
+{
+    fn view_npy(buf: &'a [u8]) -> Result<Self, ReadNpyError> {
+        let mut reader = buf;
+        let header = Header::from_reader(&mut reader)?;
+        let shape = header.shape.into_dimension();
+        let ndim = shape.ndim();
+        let len = match shape.size_checked() {
+            Some(len) if len <= std::isize::MAX as usize => len,
+            _ => return Err(ReadNpyError::LengthOverflow),
+        };
+        let data = A::bytes_as_slice(&reader, &header.type_descriptor, len)?;
+        ArrayView::from_shape(shape.set_f(header.fortran_order), data)
+            .unwrap()
+            .into_dimensionality()
+            .map_err(|_| ReadNpyError::WrongNdim(D::NDIM, ndim))
+    }
+}
+
+impl<'a, A, D> ViewNpyMutExt<'a> for ArrayViewMut<'a, A, D>
+where
+    A: CastableElement,
+    D: Dimension,
+{
+    fn view_npy_mut(buf: &'a mut [u8]) -> Result<Self, ReadNpyError> {
+        let mut reader = buf as &[u8];
+        let header = Header::from_reader(&mut reader)?;
+        let shape = header.shape.into_dimension();
+        let ndim = shape.ndim();
+        let len = match shape.size_checked() {
+            Some(len) if len <= std::isize::MAX as usize => len,
+            _ => return Err(ReadNpyError::LengthOverflow),
+        };
+        let mid = buf.len() - reader.len();
+        let data = A::bytes_as_slice_mut(&mut buf[mid..], &header.type_descriptor, len)?;
+        ArrayViewMut::from_shape(shape.set_f(header.fortran_order), data)
+            .unwrap()
+            .into_dimensionality()
+            .map_err(|_| ReadNpyError::WrongNdim(D::NDIM, ndim))
+    }
+}
+
+/// An array element type that can be cast from an `.npy` or uncompressed `.npz` file.
+pub trait CastableElement: Sized {
+    /// Casts `bytes` as slice of elements of length `len`.
+    ///
+    /// This method should invoke `slice_len()` before casting.
+    fn bytes_as_slice<'a>(
+        bytes: &'a [u8],
+        type_desc: &PyValue,
+        len: usize,
+    ) -> Result<&'a [Self], ReadDataError>;
+
+    /// Casts `bytes` as mutable slice of elements of length `len`.
+    ///
+    /// This method should invoke `slice_len()` before casting.
+    fn bytes_as_slice_mut<'a>(
+        bytes: &'a mut [u8],
+        type_desc: &PyValue,
+        len: usize,
+    ) -> Result<&'a mut [Self], ReadDataError>;
+
+    /// Verifies lengths before computing slice length.
+    ///
+    /// Ensures:
+    ///
+    ///   * `data_len` equals `len` extracted from header.
+    ///   * `data_len` is an integral multiple of its element length.
+    fn slice_len<'a>(
```

Comment:
This implementation isn't quite right. Based on the way the function is used, `data_len` is the length of the data in bytes and `len` is the desired number of elements in the cast slice, but the implementation seems to be treating `len` as the desired number of bytes in the cast slice.

I'd make it a standalone function and change it to something like the following:

```rust
/// Returns `Ok(_)` iff a slice containing `bytes_len` bytes is the correct length to cast to
/// a slice with element type `T` and length `len`.
fn check_bytes_len<T>(bytes_len: usize, len: usize) -> Result<(), ReadDataError> {
    let needed_bytes = len
        .checked_mul(mem::size_of::<T>())
        .ok_or(ReadDataError::LengthOverflow)?;
    if bytes_len < needed_bytes {
        Err(ReadDataError::MissingData)
    } else if bytes_len > needed_bytes {
        Err(ReadDataError::ExtraBytes(bytes_len - needed_bytes))
    } else {
        Ok(())
    }
}
```

Then, the cast implementation would become this:

```rust
            fn bytes_as_slice_mut<'a>(
                bytes: &'a mut [u8],
                type_desc: &PyValue,
                len: usize,
            ) -> Result<&'a mut [Self], ReadDataError> {
                match *type_desc {
                    PyValue::String(ref s) if s == $native_desc =>
                        check_bytes_len::<Self>(bytes.len(), len)?;
                        unsafe {
                            std::slice::from_raw_parts_mut(bytes.as_mut_ptr().cast(), len)
                        },
```

Reply:
Your right. I somehow assumed `len` (extracted from header) is in bytes but it is the number of elements. Then this "no integral multiple"-case doesn't exist, it's just `MissingData`.